quil-rs/src/instruction/extern_call.rs

```diff
@@ -0,0 +1,2060 @@
+        Ok(())
+    }
+
+    /// Resolve the [`Call`] instruction to any of the given [`ExternSignature`] and memory regions.
+    /// If no matching extern instruction is found, return an error.
+    pub fn resolve(
+        &mut self,
+        memory_regions: &IndexMap<String, MemoryRegion>,
+        extern_definitions: &[ExternDefinition],
+    ) -> Result<(), CallResolutionError> {
+        for definition in extern_definitions {
+            if definition.name == self.name {
+                let signature = definition
+                    .signature
+                    .as_ref()
+                    .ok_or_else(|| CallResolutionError::NoSignature(self.name.clone()))?;
+                return self
+                    .resolve_to_signature(signature, memory_regions)
+                    .map_err(|error| CallResolutionError::Signature {
+                        name: self.name.clone(),
+                        error,
+                    });
+            }
+        }
+        Err(CallResolutionError::NoMatchingExternInstruction(
+            self.name.clone(),
+        ))
+    }
+}
+
+impl Quil for Call {
+    fn write(
+        &self,
+        f: &mut impl std::fmt::Write,
+        fall_back_to_debug: bool,
+    ) -> crate::quil::ToQuilResult<()> {
+        write!(f, "CALL {}", self.name)?;
+        self.arguments.write(f, fall_back_to_debug)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rstest::*;
+
+    /// Test cases for the `ExternDefinition` Quil representation.
+    struct ExternDefinitionQuilTestCase {
+        /// The extern definition to test.
+        definition: ExternDefinition,
+        /// The expected Quil representation.
+        expected: &'static str,
+    }
+
+    impl ExternDefinitionQuilTestCase {
+        /// Signature with return and parameters
+        fn case_01() -> Self {
+            Self {
+                definition: ExternDefinition {
+                    name: "foo".to_string(),
+                    signature: Some(ExternSignature {
+                        return_type: Some(ScalarType::Integer),
+                        parameters: vec![
+                            ExternParameter {
+                                name: "bar".to_string(),
+                                mutable: false,
+                                data_type: ExternParameterType::Scalar(ScalarType::Integer),
+                            },
+                            ExternParameter {
+                                name: "baz".to_string(),
+                                mutable: true,
+                                data_type: ExternParameterType::FixedLengthVector(Vector {
+                                    data_type: ScalarType::Bit,
+                                    length: 2,
+                                }),
+                            },
+                        ],
+                    }),
+                },
+                expected: "foo \"INTEGER (bar : INTEGER, baz : mut BIT[2])\"",
```

Comment:
Raw quotes might be easier to read, here and below:
```suggestion
                expected: r#"foo "INTEGER (bar : INTEGER, baz : mut BIT[2])""#,
```